### DISCUSSION TITLE
response_model with Custom format

### DISCUSSION THREAD

**Discussion by louishot** [fastapi/fastapi#8797] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class SourceLinkResponse(BaseModel):
    id: int
    name: Optional[str]
    url: Optional[str]
    source: Optional[str]
    task_status: Optional[str]
    task_reason: Optional[str]
    updated_at: Optional[datetime]
    created_at: Optional[datetime]
    class Config:
        orm_mode = True

router = APIRouter(
    prefix="/admin"
)


@router.post("/SourceLink", response_model=schemas.SourceLinkResponse)
def Create_SourceLink(
    data: schemas.SourceLinkCreate, db: Session = Depends(get_db)
):
    db_source_link = crud.get_source_link_by_url(db, data.url)
    if db_source_link:
        raise CustomHTTPException(err_code=102, errors=[{"url": "already exists"}], messages="already exists")
    return crud.create_source_link(db=db, data=data)


output
{
    "id": 20,
    "name": "string",
    "url": "string24",
    "source": "string",
    "task_status": "string",
    "task_reason": "string",
    "updated_at": "2021-08-20T23:26:19.731234",
    "created_at": "2021-08-20T23:26:19.731234"
}

But I hope to get the following standard format

{
   "success":true,
   "err_code":0,
   "errors":[
      
   ],
   "messages":"",
   "result":[
      {
         "id":20,
         "name":"string",
         "url":"string24",
         "source":"string",
         "task_status":"string",
         "task_reason":"string",
         "updated_at":"2021-08-20T23:26:19.731234",
         "created_at":"2021-08-20T23:26:19.731234"
      }
   ]
}
```

### Description

if use response_class it will skip response_model.

Thank you

### Operating System

macOS

### Operating System Details

Macos

### FastAPI Version

0.68.0

### Python Version

3.9

**Comment by ml-evs** [comment]
Hi, you can define a separate response model that encapsulates your data model `SourceLinkResponse` in the way you want, something like:

```python

from typing import List
from pydantic import BaseModel, Field

class APIResponse(BaseModel):
     success: bool = Field(True)
     err_code: int = Field(0)
     messages: List[str] = Field(...)
     result: List[SourceLinkResponse] = Field(...)
```
such that you can return `APIResponse(result=[crud.create_source_link(db=db, data=data)])` in your route.


There are quite a few standards for JSON responses ([JSON:API](https://jsonapi.org/format/1.1/) etc.), and I'm not sure which one you are referring to, but there may be an existing library that has already defined the models for you.

**Comment by louishot** [comment]
> Hi, you can define a separate response model that encapsulates your data model `SourceLinkResponse` in the way you want, something like:
> 
> ```python
> from typing import List
> from pydantic import BaseModel, Field
> 
> class APIResponse(BaseModel):
>      success: bool = Field(True)
>      err_code: int = Field(0)
>      messages: List[str] = Field(...)
>      result: List[SourceLinkResponse] = Field(...)
> ```
> 
> such that you can return `APIResponse(result=[crud.create_source_link(db=db, data=data)])` in your route.
> 
> There are quite a few standards for JSON responses ([JSON:API](https://jsonapi.org/format/1.1/) etc.), and I'm not sure which one you are referring to, but there may be an existing library that has already defined the models for you.

worked.
Thank you so much 

```
class Response(BaseModel):
    id: int
    name: str
    url: str
    source: Optional[str]
    status: Optional[str]
    reason: Optional[str]
    updated_at: Optional[datetime]
    created_at: datetime

    class Config:
        orm_mode = True

class APIResponse(BaseModel):
    success: bool = True
    err_code: int = 0
    errors: list = []
    messages: str = ""
    result: List[Response]
return core.schemas.source_link.APIResponse(result=db_data)
```

**Comment by Matfork** [comment]
Uhm, I'm also sharing my solution to you, so you can avoid creating 'n' APIResponse classes per each route:

Assuming you want this JSON response:
```json
{
  "code": 0,
  "data": {
    "email": "string",
    "password": "string",
    "name": "string",
   }
  }
  ```
  
Where 
```json
{
    "email": "string",
    "password": "string",
    "name": "string",
 }
``` 
is a Pydantic Modal you defined with:
```python

class UserBase(BaseModel):
    email: str
    password: str
    name: str
```

Then you can define your route like this:
```python
@app.post("/{id}", responses={**getResponses('addUser', UserBase)})
async def addUser():
  pass
```
The magic is done in getResponses which is a method that generate dynamic pydantic models that can be used as APIResponses:
```python
from typing import Any
from pydantic.main import create_model

def getResponses(route: str, model: Any) -> dict:
    responses: dict = {
        200: {
            "model": create_model(route, code=(int, ...), data=(model, ...))
        }
    }
    return responses
```
Notice that args of create_model are the structure I mentioned in first part.
With this approach you can reuse the method for other routes by creating api responses dynamically. e.g.

```python
@app.patch("/{id}", responses={**getResponses('updateUser', AnotherUserBase)})
async def updateUser():
  pass
```

**Comment by JhonSanz** [comment, marked as answer]
I did it like this

```python

from pydantic import BaseModel, ConfigDict

T = TypeVar("T")

class StandardResponse(BaseModel, Generic[T]):
    model_config = ConfigDict(from_attributes=True)

    ok: bool
    msg: Optional[str] = ""
    data: Optional[T] = None


class SourceLinkResponse(BaseModel):
    id: int
    name: Optional[str]
    url: Optional[str]
    source: Optional[str]
    task_status: Optional[str]
    task_reason: Optional[str]
    updated_at: Optional[datetime]
    created_at: Optional[datetime]
    class Config:
        orm_mode = True


@router.get("/payments/", response_model=StandardResponse[List[SourceLinkResponse]])
def get_example():
   data = # list of items that mach SourceLinkResponse schema
   return StandardResponse(ok=True, msg="cats are better than dogs", data=data)
```
In StandardResponse[List[`SourceLinkResponse`]] you can replace `SourceLinkResponse` with other schemas...
Therefore, you avoid creating multiple APIResponse models. It's worth it to keep the same response json schema for errors 🤔
